funding/commitment_type_negotiation.go

```diff
@@ -24,38 +18,47 @@ var (
 // negotiateCommitmentType negotiates the commitment type of a newly opened
 // channel. If a desiredChanType is provided, explicit negotiation for said type
 // will be attempted if the set of both local and remote features support it.
-// Otherwise, implicit negotiation will be attempted. Two booleans are
-// returned letting the caller know if the option-scid-alias or zero-conf
-// channel types were negotiated.
+// Otherwise, implicit negotiation will be attempted.
+//
+// The returned ChannelType is nil when implicit negotiation is used. An error
+// is only returned if desiredChanType is not supported.
 func negotiateCommitmentType(desiredChanType *lnwire.ChannelType, local,
-	remote *lnwire.FeatureVector, mustBeExplicit bool) (bool,
-	*lnwire.ChannelType, lnwallet.CommitmentType, error) {
-
-	if desiredChanType != nil {
-		// If the peer does know explicit negotiation, let's attempt
-		// that now.
-		if hasFeatures(
-			local, remote, lnwire.ExplicitChannelTypeOptional,
-		) {
+	remote *lnwire.FeatureVector) (*lnwire.ChannelType,
+	lnwallet.CommitmentType, error) {
 
+	// BOLT#2 specifies we MUST use explicit negotiation if both peers
+	// signal for it.
+	if hasFeatures(local, remote, lnwire.ExplicitChannelTypeOptional) {
+		if desiredChanType != nil {
 			commitType, err := explicitNegotiateCommitmentType(
 				*desiredChanType, local, remote,
 			)
-			return true, desiredChanType, commitType, err
-		}
 
-		// If we're the funder, and we are attempting to use an
-		// explicit channel type, but the remote party doesn't signal
-		// the bit, then we actually want to exit here, to ensure the
-		// user doesn't end up with an unexpected channel type via
-		// implicit negotiation.
-		if mustBeExplicit {
-			return false, nil, 0, errUnsupportedExplicitNegotiation
+			return desiredChanType, commitType, err
 		}
+
+		// Explicitly signal the "implicit" negotiation commitment type
+		// as default when a desired channel type isn't specified.
+		chanType, commitType := implicitNegotiateCommitmentType(local,
+			remote)
+
+		return chanType, commitType, nil
 	}
 
+	// Otherwise, we'll use implicit negotiation. In this case, we are
```

Comment:
What if `ExplicitChannelTypeRequired` is set?

Reply:
Good question... it appears this is a case not handled by the existing code either.  I guess none of the lightning implementations actually signal this feature bit yet?

Probably outside the scope of this PR, but it seems like something we should definitely create an issue for and follow up.

Reply:
If one node has `ExplicitChannelTypeRequired` and the other does not support `ExplicitChannelType` at all, they'll just hang up on `Init` as they [don't speak each others languages](https://github.com/carlaKC/lightning-rfc/blob/master/01-messaging.md#the-init-message): 

```
- upon receiving unknown _even_ feature bits that are non-zero:
    - MUST close the connection.
```

Reply:
Yep! `hasFeature` checks both, was just pointing out that we don't need to spin up an issue, per: 

> ok I think hasFeatures will implicitly check for this bit...def worth an issue.

I probably misunderstood the original comment - nothing further here from me!